### PR TITLE
Simplify NAT46x64,recorder tests

### PR DESCRIPTION
**Pull request by joestringer**
Requires https://github.com/cilium/cilium/pull/32070 to test.

Review commit-by-commit to reason through the changes over the course of the PR.

Previously, these tests would emit very verbose output which was hard to
interpret when it failed, for instance a test may fail and this is the output:

```
++[01:02:57] docker exec -t lb-node ip -o -4 a s eth0
++[01:02:57] awk '{print $4}'
++[01:02:57] head -n1
++[01:02:57] cut -d/ -f1
+[01:02:57] LB_NODE_IP=172.12.42.2
+[01:02:57] ip r a 10.0.0.8/32 via 172.12.42.2
++[01:02:57] seq 1 10
+[01:02:57] for i in $(seq 1 10)
+[01:02:57] curl -s -o /dev/null '[fd00:cafe::1]:80'
+[01:02:57] for i in $(seq 1 10)
+[01:02:57] curl -s -o /dev/null '[fd00:cafe::1]:80'
+[01:02:57] for i in $(seq 1 10)
+[01:02:57] curl -s -o /dev/null '[fd00:cafe::1]:80'
+[01:02:57] for i in $(seq 1 10)
+[01:02:57] curl -s -o /dev/null '[fd00:cafe::1]:80'
+[01:02:57] for i in $(seq 1 10)
+[01:02:57] curl -s -o /dev/null '[fd00:cafe::1]:80'
+[01:02:57] for i in $(seq 1 10)
+[01:02:57] curl -s -o /dev/null '[fd00:cafe::1]:80'
+[01:02:57] for i in $(seq 1 10)
+[01:02:57] curl -s -o /dev/null '[fd00:cafe::1]:80'
+[01:02:57] for i in $(seq 1 10)
+[01:02:57] curl -s -o /dev/null '[fd00:cafe::1]:80'
+[01:02:57] for i in $(seq 1 10)
+[01:02:57] curl -s -o /dev/null '[fd00:cafe::1]:80'
+[01:02:57] for i in $(seq 1 10)
+[01:02:57] curl -s -o /dev/null '[fd00:cafe::1]:80'
++[01:02:57] seq 1 10
+[01:02:57] for i in $(seq 1 10)
+[01:02:57] curl -s -o /dev/null 10.0.0.8:80
+[01:02:57] echo 'Failed 1'
Failed 1
+[01:02:57] exit -1
Error: Process completed with exit code 255.
```

If you understand what the test is trying to do and where in the test that
these commands were run, then you might be able to debug it. But for most
mortals, this output was not very easy to understand.

With this PR, the tests are reworked to only print verbose output when the
environment variable `V=1` is set, and by default it prints a much terser
set of output, such as:

```
[./test/nat46x64/test.sh:228]     Installing Cilium with Mode:TC  Algorithm:Maglev        Recorder:Disabled
cilium-lb
9c536a08a4bdb8751cdc2337503c5864e1fd1984619635b466cf4352d99d71b2
cilium: daemon unreachable
cilium: daemon unreachable
OK
[./test/nat46x64/test.sh:229]     Testing service [fd00:cafe::1]:80 -> 172.12.42.3:80 via TC + Maglev
[./test/nat46x64/test.sh:233]     Installing Cilium with Mode:TC  Algorithm:Random        Recorder:Disabled
cilium-lb
894298c1821b2b536f6744c78aa886bc8f3ab7f4a291b82d9ccb6677d253f26b
cilium: daemon unreachable
cilium: daemon unreachable
OK
[./test/nat46x64/test.sh:234]     Testing service [fd00:cafe::1]:80 -> 172.12.42.3:80 via TC + Random
[./test/nat46x64/test.sh:238]     Configuring service 10.0.0.8:80 -> 172.12.42.3:80
Deprecation warning: --id parameter will change from int to string in v1.14
Creating new service with id '2'
Added service with 1 backends
[./test/nat46x64/test.sh:244]     Installing route: 'ip -4 r a 10.0.0.8/32 via 172.12.42.2'
[./test/nat46x64/test.sh:246]     Checking connectivity via [fd00:cafe::1]:80 -> 172.12.42.3:80
[./test/nat46x64/test.sh:249]     Checking connectivity via 10.0.0.8:80 -> 172.12.42.3:80
[./test/nat46x64/test.sh:252]     Installing Cilium with Mode:TC  Algorithm:Maglev        Recorder:Disabled
cilium-lb
89f2ad155a0a5ae0922ba741baa8d6186d4f87ee4a3e019d5845a1847b6f5f1d
cilium: daemon unreachable
cilium: daemon unreachable
OK
[./test/nat46x64/test.sh:253]     Testing service [fd00:cafe::1]:80 -> 172.12.42.3:80 via TC + Maglev
Deprecation warning: --id parameter will change from int to string in v1.14
Service 1 deleted successfully
Deprecation warning: --id parameter will change from int to string in v1.14
Service 2 deleted successfully
```

Or, on failure:

```
$ sudo ./test/nat46x64/test.sh
[./test/nat46x64/test.sh:197]     Installing Cilium with Mode:TC  Algorithm:Maglev        Recorder:Disabled
cilium-lb
ccb3d8f50c6d15a016035eb81c12e62efa04ef79cad62a00a1c1b9bd529ca58c
cilium: daemon unreachable
OK
[./test/nat46x64/test.sh:199]     Configuring service 10.0.0.4:80 -> [2001:db8:1::3]:80
Deprecation warning: --id parameter will change from int to string in v1.14
Creating new service with id '1'
Added service with 1 backends
[./test/nat46x64/test.sh:209]     Testing service 10.0.0.4:80 -> [2001:db8:1::3]:80 via TC + Maglev
[./test/nat46x64/test.sh:210]     Installing route: 'ip -4 r a 10.0.0.4/32 via 172.12.42.2'
[./test/nat46x64/test.sh:211]     Failed connection from localhost to 10.0.0.4:80 (attempt 5/10)
```

The annotated lines in the square brackets also provide a test file and line number where the failure occurred to make it easier to track down the failure and attempt to reproduce.

### CODE REVIEW

File: test/nat46x64/test.sh [hunk 1]
```diff
@@ -1,423 +1,411 @@
 #!/usr/bin/env bash
 
 PS4='+[\t] '
```
Comment (joestringer) [inline test/nat46x64/test.sh:3]: So heads up by default these timestamps won't show up when you run the script, unless you set the env var `V=1` like `V=1 sudo -E ./test/nat46x64/test.sh`. I think this is fine as a tradeoff, in GitHub actions we have access to timestamps anyway through the UI:

image

File: test/nat46x64/test.sh [hunk 2]
```diff
@@ -1,423 +1,411 @@
...
+fi
+
+info "Checking the list of filters in IPv6 recorder ID:2"
+if [ $(${CILIUM_EXEC} cilium-dbg bpf recorder list | grep "ID:2" | wc -l) -ne ${#RECORDER_FILTERS_IPV6[@]} ]; then
+    echo "Expected filters:"
+    echo "${RECORDER_FILTERS_IPV6[@]}" | nl -bn
+    echo "Found filters:"
+    ${CILIUM_EXEC} cilium-dbg bpf recorder list | nl -bn
+    fatal "Recorder filters did not match expected list"
+fi
```
Comment (tklauser) [inline test/nat46x64/test.sh]: nit: these two blocks look similar enough that they might be refactored into a function as well.

**Review by tklauser — APPROVED** [review]
Thanks for splitting this up into nicely documented commits (took a while though to read through all of them :wink:). One non-blocking nit about a potential refactoring inline.

File: test/nat46x64/test.sh [hunk 3]
```diff
@@ -53,12 +53,11 @@ cilium_install \
 IFIDX=$(docker exec -i lb-node \
     /bin/sh -c 'echo $(( $(ip -o l show eth0 | awk "{print $1}" | cut -d: -f1) ))')
 LB_VETH_HOST=$(ip -o l | grep "if$IFIDX" | awk '{print $2}' | cut -d@ -f1)
-ethtool -K $LB_VETH_HOST rx off tx off
+ethtool -K "$LB_VETH_HOST" rx off tx off
 
 NGINX_PID=$(docker inspect nginx -f '{{ .State.Pid }}')
-WORKER_IP4=$(nsenter -t $NGINX_PID -n ip -o -4 a s eth0 | awk '{print $4}' | cut -d/ -f1 | head -n1)
-WORKER_IP6=$(nsenter -t $NGINX_PID -n ip -o -6 a s eth0 | awk '{print $4}' | cut -d/ -f1 | head -n1)
-WORKER_MAC=$(nsenter -t $NGINX_PID -n ip -o l show dev eth0 | grep -oP '(?<=link/ether )[^ ]+')
```
Comment (aspsk) [inline test/nat46x64/test.sh:61]: Had this line caused a warning? Or just an unquoted variable?

File: test/nat46x64/test.sh [hunk 4]
```diff
@@ -60,7 +60,30 @@ function initialize_docker_env {
     ethtool -K "$LB_VETH_HOST" rx off tx off
 }
 
+function force_cleanup {
+    ${CILIUM_EXEC} cilium-dbg service delete 1 || true
```
Comment (aspsk) [inline test/nat46x64/test.sh:125]: I usually use `smth || :` instead :)

File: test/nat46x64/test.sh [hunk 5]
```diff
@@ -80,6 +80,17 @@ function cleanup {
     force_cleanup
 }
 
+# $1 - target service, for example "[fd00:dead:beef:15:bad::1]:80"
+function wait_service_ready {
+    # Try and sleep until the LB2 comes up, seems to be no other way to detect when the service is ready.
+    set +e
+    for i in $(seq 1 10); do
+        curl -s -o /dev/null "${LB_ALT}" && break
```
Comment (aspsk) [inline test/nat46x64/test.sh]: the `URL` is passed via a function parameter, however, "${LB_ALT}" (without :80) is hardcoded here. Should be `$1` instead or
```
local url="$1"
...
curl ... $url
```

**Review by aspsk — COMMENTED** [review]
LGTM, see a few minor comments

**Comment by ti-mo** [comment]
@aspsk Try to 'Request Changes' when leaving a review, otherwise I would've merged this as-is. I'll remove the ready label for now.

**Review by aspsk — CHANGES_REQUESTED** [review]
Hereby I am marking my previous review as "request for changes"

File: test/nat46x64/test.sh (hunk 3 above)
Comment (joestringer) [reply test/nat46x64/test.sh:61]: I think `WORKER_MAC` was entirely unused. Otherwise the `WORKER_IP*` vars had an unquoted `$NGINX_PID` yeah.

Comment (joestringer) [reply test/nat46x64/test.sh:61]: If your question was in reference to a particular commit, I could take a deeper look at that commit and the context to make sure that individual change didn't miss something.

File: test/nat46x64/test.sh (hunk 4 above)
Comment (joestringer) [reply test/nat46x64/test.sh:125]: TIL. Tidier but harder to read if you're less familiar with bash, which applies to most contributors in the codebase.

File: test/nat46x64/test.sh (hunk 5 above)
Comment (joestringer) [reply test/nat46x64/test.sh]: Ah, good find, let me fix that up.

File: test/nat46x64/test.sh (hunk 2 above)
Comment (joestringer) [reply test/nat46x64/test.sh]: I agree. I did try this earlier when developing these tests, but being able to properly pass the array contents into a function was a little tricky. I can try to take another stab at this, but if it's not trivial then I'll probably skip this feedback. IMO we're reaching diminishing returns on iterating on these steps at this point.

Recorder is also a bit weird to be combined into the nat46x64 test, it seems to be entirely unrelated; I wouldn't mind if it were in its own file. But a bit more refactoring is necessary to get that done. We can continue to iterate in the tree.

Comment (joestringer) [reply test/nat46x64/test.sh]: Looks easy enough. I'll add an extra commit on top:

```
diff --git a/test/nat46x64/test.sh b/test/nat46x64/test.sh
index 6366f448a838..75a5aa0af4e0 100755
--- a/test/nat46x64/test.sh
+++ b/test/nat46x64/test.sh
@@ -258,6 +258,23 @@ function test_services {
     ${CILIUM_EXEC} cilium-dbg service delete 2
 }

+# Check whether the list of filters matches the expected list.
+#
+# $1 - ID
+# $@ - Filters
+function check_recorder_list {
+    ID=$1
+    shift
+    num_filters=$#
+    if [ $(${CILIUM_EXEC} cilium-dbg bpf recorder list | grep "ID:${ID}" | wc -l) -ne "$num_filters" ]; then
+        echo "Expected filters:"
+        echo "$@" | nl -bn
+        echo "Found filters:"
+        ${CILIUM_EXEC} cilium-dbg bpf recorder list | nl -bn
+        fatal "Recorder filters did not match expected list"
+    fi
+}
+
 force_cleanup 2>&1 >/dev/null
 initialize_docker_env
 trap cleanup EXIT
@@ -383,23 +400,8 @@ ${CILIUM_EXEC} \
     cilium-dbg recorder update --id 2 --caplen 100 \
         --filters="$(printf '%s,' "${RECORDER_FILTERS_IPV6[@]}" | sed 's/,*$//')"

-info "Checking the list of filters in IPv4 recorder ID:1"
-if [ $(${CILIUM_EXEC} cilium-dbg bpf recorder list | grep "ID:1" | wc -l) -ne ${#RECORDER_FILTERS_IPV4[@]} ]; then
-    echo "Expected filters:"
-    echo "${RECORDER_FILTERS_IPV4[@]}" | nl -bn
-    echo "Found filters:"
-    ${CILIUM_EXEC} cilium-dbg bpf recorder list | nl -bn
-    fatal "Recorder filters did not match expected list"
-fi
-
-info "Checking the list of filters in IPv6 recorder ID:2"
-if [ $(${CILIUM_EXEC} cilium-dbg bpf recorder list | grep "ID:2" | wc -l) -ne ${#RECORDER_FILTERS_IPV6[@]} ]; then
-    echo "Expected filters:"
-    echo "${RECORDER_FILTERS_IPV6[@]}" | nl -bn
-    echo "Found filters:"
-    ${CILIUM_EXEC} cilium-dbg bpf recorder list | nl -bn
-    fatal "Recorder filters did not match expected list"
-fi
+check_recorder_list 1 "${RECORDER_FILTERS_IPV4[@]}"
+check_recorder_list 2 "${RECORDER_FILTERS_IPV6[@]}"

 trace_exec ${CILIUM_EXEC} cilium-dbg recorder list
 trace_exec ${CILIUM_EXEC} cilium-dbg bpf recorder list
```

**Review by aspsk — APPROVED** [review]

**Pull request merged**